src/attached_probe.h

```diff
@@ -40,84 +40,42 @@ class AttachedProbe {
+  friend class AttachedProbe;
+};
+
+class AttachedWatchpointProbe : public AttachedProbe {
+public:
+  ~AttachedWatchpointProbe() override = default;
+
+  const uint64_t probe_addr;
 
 private:
-  AttachedProbe(Probe &probe,
-                const BpfProgram &prog,
-                std::optional<int> pid,
-                BPFtrace &bpftrace,
-                bool safe_mode);
-  Result<> attach();
-  std::string eventprefix() const;
-  std::string eventname() const;
-  Result<uint64_t> resolve_offset(const std::string &path,
-                                  const std::string &symbol,
-                                  uint64_t loc);
-  Result<> resolve_offset_kprobe();
-  Result<> resolve_offset_uprobe(bool safe_mode);
-  Result<> resolve_offset_uprobe_multi(const std::string &path,
-                                       const std::string &probe_name,
-                                       const std::vector<std::string> &funcs,
-                                       std::vector<std::string> &syms,
-                                       std::vector<unsigned long> &offsets);
-  Result<> attach_multi_kprobe();
-  Result<> attach_multi_uprobe();
-  Result<> attach_kprobe();
-  Result<> attach_uprobe(bool safe_mode);
-
-  // Note: the following usdt attachment functions will only activate a
-  // semaphore if one exists.
-  //
-  // Increment semaphore count manually with memory hogging API (least
-  // preferable)
-  int usdt_sem_up_manual(const std::string &fn_name, void *ctx);
-  // Increment semaphore count manually with BCC addsem API
-  int usdt_sem_up_manual_addsem(int pid, const std::string &fn_name, void *ctx);
-  int usdt_sem_up(BPFfeature &feature,
-                  int pid,
-                  const std::string &fn_name,
-                  void *ctx);
-  Result<> attach_usdt(BPFfeature &feature);
-
-  Result<> attach_tracepoint();
-  Result<> attach_profile();
-  Result<> attach_interval();
-  Result<> attach_software();
-  Result<> attach_hardware();
-  Result<> attach_watchpoint(const std::string &mode);
-  Result<> attach_fentry();
-  int detach_fentry();
-  Result<> attach_iter();
-  int detach_iter();
-  Result<> attach_raw_tracepoint();
-  int detach_raw_tracepoint();
-
-  Result<> check_alignment(std::string &path,
-                           std::string &symbol,
-                           uint64_t sym_offset,
-                           uint64_t func_offset,
-                           bool safe_mode,
-                           ProbeType type);
-
-  Probe &probe_;
-  const BpfProgram &prog_;
-  BPFtrace &bpftrace_;
-  std::optional<int> pid_;
-  bool safe_mode_;
-
-  std::vector<int> perf_event_fds_;
-  uint64_t offset_ = 0;
-  int tracing_fd_ = -1;
-  std::function<void()> usdt_destructor_;
-  USDTHelper usdt_helper;
+  AttachedWatchpointProbe(int progfd,
+                          std::function<void()> cleanup,
+                          uint64_t probe_addr);
+  friend class AttachedProbe;
```

Comment:
I'm not a fan of the use of `friend` here - to me it indicates the class hierarchy is not properly designed. As an alternative perhaps consider `protected` members? But see my overall comment as well